### ISSUE TITLE
ODROID-N2 doesn't boot from Kingston eMMC modules

### ISSUE THREAD

**Issue by piercy** [home-assistant/operating-system#3250]
### Describe the issue you are experiencing

I recently purchased an odroid n2+ and 32gb emmc module.  After flashing `haos_odroid-n2-12.1.img.xz`, first with a usb emmc device and then trying to do it via USB connection with the odroid n2+ device.  The flash was successful in both cases, but once the device was powered up I couldnt access HAOS via IP or homeassistant.local.   I knew the IP was correct as I could compare the mac address.

There was no blue light on the board, just a red one.  After some trial and error I went back to HAOS 12.0 and it flashed and worked first time. I had a blue light and I could immediately access it through a browser.

### What operating system image do you use?

odroid-n2 (Hardkernel ODROID-N2/N2+)

### What version of Home Assistant Operating System is installed?

12.1

### Did you upgrade the Operating System.

No

### Steps to reproduce the issue

1. Flash `haos_odroid-n2-12.1.img.xz` to an odroid device
2. Power it on
3. observe no blue light and home assistant is inaccessible
...

### Anything in the Supervisor logs that might be useful for us?

```txt
Couldn't get to logs as it wouldn't boot
```

### Anything in the Host logs that might be useful for us?

```txt
Couldn't get to logs as it wouldn't boot
```

**Comment by scoobydrvr** [comment]
Just jumping in to say I have a very similar issue with my new n2+ as well...

https://github.com/home-assistant/operating-system/issues/3227

**Comment by piercy** [comment]
I have been struggling with this again today, seems to not be based on the version. neither 12.0 or 12.1 work for me.  However, if i flash ubuntu the device boots fine. So i dont think it can be a device or emmc issue.  

No idea what to do any more but it seems like I cant use HAOS with my odroid n2+

**Comment by scoobydrvr** [comment]
Just FYI - I had to buy a USB to serial converter to get any indication of what's going on. I've been informed they changed the bootloader between 11.4 and 11.5. 11.4 boots fine but 11.5 gives me a message over serial saying that it's unable to find the eMMC card (see my thread above for specifics).

**Comment by piercy** [comment]
> Just FYI - I had to buy a USB to serial converter to get any indication of what's going on. I've been informed they changed the bootloader between 11.4 and 11.5. 11.4 boots fine but 11.5 gives me a message over serial saying that it's unable to find the eMMC card (see my thread above for specifics).

Yeah, im following that now too, unfortunately 11.4 doesnt work for me either.

No idea what to do any more, on yours i noticed a successful boot seems to use mmc 1 but unsucessful is trying mmc 2. No idea why that would be but im at a loss on what else to do any more.

I basically just cant use HAOS at the moment, even though i had it running for 3 days on 12.0 before it died... i have no idea how to even begin trying to solve this

**Comment by piercy** [comment]
I tried 12.2.rc1 today, the issue still persists

**Comment by agners** [comment]
Let's track it in #3227 as this seems to be the issue with most information at this point.

**Comment by sairon** [comment]
Reopening this after discoveries from https://github.com/home-assistant/operating-system/issues/3227#issuecomment-2175983018. This issue is clearly a problem with the Kingston eMMC modules which never worked with the upstream U-Boot and the 24 MHz we carry to mitigate some other eMMC-related issues. Since we pulled back the 12.3 release for ODROID N2 which was the only one that fixed booting with the Kingston eMMC modules, it was actually never resolved.

**Comment by piercy** [comment]
Regarding version, I haven't changed my HAOS versions since installation. So I would assume its still on 12.3.  I have been updating HA Core but I imagine that doesn't affect this.

**Comment by piercy** [comment]
So I didn't upgrade, in fact deliberately skipped the upgrade. And on a reboot it now wont boot again. Literally so sick of this, first trip into home assistant, ordroid and MMCs and I end with a PoS that barely works. 

sorry for the rant but its so frustrating how anyone is supposed to know that their MMC is going to be an issue.

**Comment by piercy** [comment]
Just a further update. I have decided to no longer use HA. This was my first time using HA. **This experience has been utterly awful and left an incredibly sour taste in my mouth.** I am the unlucky person to buy a "supported" - or at least it wasn't unsupported - EMMC, and then have the developers unsupport it on me.   I don't believe I have received much help regarding this issue, or getting it to work. The small amount of help I did receive only ended up in the EMMC being unsupported.  Which is basically telling me to get stuffed and deal with it.  I realise this is open source, and there's no reason for anyone to provide support, but there is also an entire company behind it but yet here I am struggling.

There only seems to be one company available for me to buy an EMMC from, and they cannot guarantee to provide me with a non-Kingston emmc. So i am basically screwed until someone with knowledge decides to pick this up and fix it.

I have given up on HAOS, and as a first experience into it, its been utterly terrible. 

I thank anyone who has tried, and sorry this may come across as ungrateful but please understand i've spent 100s of pounds on supported hardware only to me told, yeah your not supported any more - Good luck on your own.

**Comment by amiler** [comment]
Kingston 32GB device does not work on multiple N2+ systems. I have tried every single HAOS version from 11.3 up to the latest 13.1. None worked. Device tries to boot from network with "serverIP not set" console message. Just hangs. Other OS (Ubuntu) flashed on this 32GB eMMC does work without problems.
I have 64GB device (not Kingston) with same revision pcb that does boot with any HAOS version. Unfortunately, I cannot clone 64GB and use on 32GB. Is there anything that eMMC "driver" owner(s) is(are) going to do about this or we have to start getting non Kingston eMMC modules from now on?

Pics attached of working and non-working eMMC device.
![eMMC](https://github.com/user-attachments/assets/1717e4ea-aaa4-450e-bbec-e673f9e7fc6b)

**Comment by yfrans** [comment]
Same here.
New ODROID N2+, HAOS v13.1, no boot.
Tried HAOS v10.1 and it worked.

Screenshot from non working v13.1:
![image](https://github.com/user-attachments/assets/144f3d18-ed78-449b-9e33-10f25ca4d01e)

**Comment by piercy** [comment]
In a strange, turn of events, I am currently running 13.1 on my 32gb kingston emmc.   I am not restarting it, for fear that it wont turn back on.. however, it does seem very strange that I am now working fine on 13.1. 

Posting this mainly to maybe help the developers here identify what may be going on.  It doesn't seem as simple as it doesn't work with kingston emmc's, but something a little more intricate as it works in some cases.

**Comment by sairon** [comment]
Sorry for leaving you hanging here for such a long time, and sorry @piercy for the bad experience you had with Home Assistant. Unfortunately, this is something that struck us unprepared and it took a while to discover what was the cause of the sudden issues on N2. It also wasn't straightforward to obtain these eMMCs, as the distributors haven't had them or refused to check what model they have in stock, which prolonged the time before we could move forward.

The problem with N2 is that the eMMC driver in U-Boot is known to have issues for quite a long time. There is a [hacky patch](https://github.com/home-assistant/operating-system/blob/6d77355c1c3f35fdb5c43cc762b224b19767cce2/buildroot-external/board/hardkernel/patches/uboot/0001-HACK-mmc-meson-gx-limit-to-24MHz.patch) for quite a while in HAOS that helped to fix the intermittent issues on the previous hardware. It's not only problem of HAOS but of other distributions using mainstream U-Boot on meson-g12b (Amlogic S922X) in general - only just a few months ago [Armbian included it too](https://forum.armbian.com/topic/36354-odroid-n2-with-armbian-bookworm-emmc-boot-fail/). The downside of the patch is that it breaks communication with new Kingstom eMMC modules, which fail (intermittently, as you noticed) to operate at the lower frequency (although they should work fine in this setting too). On the other hand, when we adjusted the patch to fix Kingston issues, it triggered regression on hardware that worked well up until that point, which is more serious than something not working "out of the box".

The nature of this hardware bug is most probably in the clock timing, and if you don't know what you're doing, you can only shoot in the dark and try fiddling with numbers, hoping it will change. To reliably analyze the issue, you need specialized hardware that can help you see what's going on on the eMMC lines and why it sometimes fails. Even though Nabu Casa is behind the Home Assitant, it doesn't mean we have ways to solve every issue that is out there, especially if it requires specialized knowledge and other resources. At this point, only telling you that there is a bug that we can't resolve and the hardware is unsupported (yet) is the only "support" we can provide you.

There are still workarounds though - you don't need to boot from an eMMC - you can use an SD card for the boot. You still might be able to use the Kingston eMMC as the data disk (which is where the most data is written), since the Linux eMMC driver doesn't seem to have the same issues as the U-Boot's does. You can also stay at HAOS 12.3, or use the U-Boot binary from this version to run any other HAOS version (I'll be glad to provide instructions how to do so) - you just need to make sure you don't upgrade the OS before the U-Boot bug is resolved - if you do, you'll have to replace the U-Boot binary again. Also, it looks like that the Kingston eMMCs only come in 32 GB size variants, that can serve as a buying guide to avoid these issues.

For some good news, I tried to find someone who could help me identify why the Kingston eMMC behaves differently, and potentially what's the difference in driving the eMMC from U-Boot and from the kernel. There is something in the works but it would not be sooner than by the end of the September though, so patience is the key now.

**Comment by mversluijs** [comment]
@sairon thnx for the update!
Have similar issues with a brand new 64GB eMMC module, probably Kingston as booting from 12.3 'fixed' it.
Looking forward to a fix.

**Comment by coconutssjaakie** [comment]
I've struggled with this issue on 32 and 64 GB Kingston eMMC. I am now running HA but I am afraid to reboot or upgrade the system. I bought myself a back power supply so my n2+ will stay alive when a power interrupt should occur.
The crazy thing is; I;ve tried to boot several versions of HA on a 64Gb eMMC. It kept me busy for a couple of days. No succes. When I returned the card to the reseller, they were able to boot it up within seconds. What the.....

**Comment by dselzle** [comment]
Fingers crossed that someone can find a solution for this, either here or upstream somewhere. I was going to order one of the new Odroid M2s until I noticed that the soldered eMMC it comes with appears to have the word "Kingston" on it. I'm going to hold off for now if booting off a microSD is the only easy option for it.

**Comment by sairon** [comment]
@dselzle The issue is only with the SoC found on ODROID N2, the Kingston eMMC works correctly on other SBCs - at least I have tested it on M1 without any issues. There is a problem though - there is no image for [ODROID M2](https://www.home-assistant.io/installation/odroid), it's out just a few weeks.

**Comment by dselzle** [comment]
> @dselzle The issue is only with the SoC found on ODROID N2, the Kingston eMMC works correctly on other SBCs - at least I have tested it on M1 without any issues. There is a problem though - there is no image for [ODROID M2](https://www.home-assistant.io/installation/odroid), it's out just a few weeks. 

Good to know, thanks!

**Comment by amiler** [comment]
I got a new eMMC module (64GB) to replace 32GB Kingston and, unfortunately, I got Kingston again. I have working system with non-Kingston eMMC that works in both N2+ systems. Comparing debug messages between the two (failing and passing boot with two different eMMC), it show that failing boot is trying to boot from MMC device at slot 2, continues to try to boot from network and fails. Both cards did not respond to "voltage select" but that did not stop non-Kingston eMMC from booting successfully. Failing boot shows an old U-Boot version because I was trying to flash an old version of HAOS (I believe it was 12.3 or older) while good boot is using the latest 13.1 version of HAOS. When HAOS is flashed onto SD card (any version), system boots without problems. 
I guess this data is not new but wanted to share in case if it helps. Hopefully, this will be resolved soon since, when bying eMMC, you cannot tell if it is Kingston or not. Clearly, issue is with Kingston eMMCs. I have tried 32GB and 64GB purchased from two different sources and they both failed.
Picture of logs (left - failed, right - successful boot)
![Failed_good_boot](https://github.com/user-attachments/assets/157441c2-b447-414e-a8ac-03236a0e890b)

**Comment by betriebsanleitung** [comment]
No help for debugging. But just the information that it seems that Hardkernel emmc modules with Kingston get more common. Bought some in Germany from different suppliers to set up a new Home Assistant on ODROID-N2+. All with Kingston.

**Comment by betriebsanleitung** [comment]
Hi @sairon, is there any news about a fix for emmcs with Kingston chip?
I tried to find an emmc without Kingston chip, but all suppliers in Europe who could check their stock told me that all emmc modules are now with Kingston. 

I bought the ODROID-N2+ because it is still recommended hardware on the Home Assistant website. And on the hassio setup page there's still no information that users with new hardware are currently unable to set up a new system as described on the website. So I guess more and more users will run into this problem. 

Is there a timeline for when we can expect a fix? I know its not easy to fix, but any information or guess is better than nothing. 
Unfortunately, I cannot return my newly purchased hardware because it took me too long to find the reason why my new system would not boot (and the reason in this thread). Now I'm wondering if it's faster to sell it on ebay and move on to other hardware than to wait for a fix. Or if a fix is planned within the next 1-3 weeks. 

Thanks for the update!

**Comment by piercy** [comment]
> I bought the ODROID-N2+ because it is still recommended hardware on the Home Assistant website.

This is a good point, HA need to stop recommending this if they aren't going to support it. The EMMC chips are getting more common. More and more people are going to get duped.

**Comment by piercy** [comment]
Also could we please remove the duplicate flag from this thread, this issue isn't a duplicate.

**Comment by sairon** [comment]
I've made some progress and things are looking good on my side, I will appreciate wider testing with the images available from [this build](https://github.com/home-assistant/operating-system/actions/runs/11279308656) (should finish in couple of hours top).

Eventually, the patch could be included in 14.0 and its pre-releases in the coming weeks. I'm hesitant to include it in the coming 13.2 as there's still a slight chance it could introduce regression on other eMMC modules.

**Comment by piercy** [comment]
@sairon Unfortunately that version doesn't boot for me.  The N2+ shows a red light for pwr but the blue sys light that usually flashes, isn't flashing (which correct me if I am wrong, usually means it has an issue and isn't trying any more).   I am also to communicate with it via the HA web interface.

There is no output on the HDMI, so I am unable to get any additional information from it.    Is there another way for me to get more information out of it?

Steps for what I did:

1. Downloaded [haos_odroid-n2-13.2.dev1728582564.img.xz](https://github.com/home-assistant/operating-system/actions/runs/11279308656/artifacts/2041952300)
2. Extracted it from the zip (it downloaded as a zip)
3. Flashed it to the emmc with balenaEtcher and let it verify, all good
4. Put the EMMC back into the N2+ and plugged the power in
5. Saw red pwr light, but no blue sys light.  For info, he switch is in the right position (so on the MMC side)

**Comment by amiler** [comment]
Patch that @sairon  put in devel branch does not work for me. Same message during boot

Card did not respond to voltage select! : -110
unable to select a mode : -5
MMC Device 2 not found
no mmc device at slot 2

Trying to source non-Kingston eMMC module.

**Comment by remconiesten** [comment]
Patch tested from the devel branch @sairon posted 3 days ago. Also tested 12.3 (didn't test this one yet, out of a number of other releases). It all ends in _*** ERROR: `serverip' not set_

This is on a 64GB Kingston eMMC, see photo. Continuing on 2-years old SD-card for now, time to replace it.

![IMG_6848](https://github.com/user-attachments/assets/3521bf7b-154e-4e63-b06c-6dcd5cb25995)

**Comment by sairon** [comment]
Thank you all for testing and reporting back, I was off by the end of the week and couldn't reply sooner. While the fix worked in isolated U-Boot tests and quick HAOS boot test, for some reason it still fails with higher rate in the normal build/initialization procedure.

> However, I power cycled it and it started booting. It was on the 13.2 dev build. However, I didn't reflash it, just power cycled and it worked. Very odd.

It's not that odd, given that if the boot fails, it doesn't reboot and try again, but tries to boot from the ethernet and then presents a U-Boot shell prompt (it's the default behavior of upstream U-Boot config and it's not documented for HAOS, so I'd like to change that as well).

Anyway, even though the bad news is that the fix doesn't work, there's still good news that I'm able to reproduce it and I still have couple of ideas what to try, just this time I'll do some more extensive/overnight testing to verify the issue is really gone. Thanks for the patience.

**Comment by piercy** [comment]
Thanks @sairon , if it helps I am piercy, on the HA discord.  If you need more immediate feedback feel free to message me.  I am happy to try random devbuilds and all sorts of stuff if it helps in the debug process.   I realize it can be painful doing a build then waiting a couple of days for people to see if it works.

 I do struggle to get debug information off of the n2+, i dont know if theres some kind of debug connector I can get to get more information from it?  Either way, if it helps I am happy to spend some time just trying stuff until we find something that works

**Comment by sairon** [comment]
Thanks @piercy for the offer, you've been very helpful with the testing here as well, next time I'll just need to do test everything more thoroughly myself first. But surely I can ping you the next time I have something to test too.

About the debug information, there's a dedicated UART connector next to the microSD card - there's either a Hardkernel's [proprietary USB-UART](https://www.hardkernel.com/shop/usb-uart-2-module-kit-copy/) module, or you can use a generic 3.3V USB-TTL converter an connect it according to the pinout documented on the [ODROID wiki](https://wiki.odroid.com/odroid-n2/application_note/gpio/uart#connect_wires). Disregard the instructions for UART_EE_A/UART_EE_B connection, U-Boot only uses the one on the CON5.

**Comment by sairon** [comment]
There's [another build](https://github.com/home-assistant/operating-system/actions/runs/11723801141) which should fix the Kingston issues, this time I hope for good, as it also passed a bit more extensive testing this time. Again, feedback from testing on other systems is welcome.

**Comment by piercy** [comment]
@sairon looks good, mine booted immediately. I am currently waiting for it initialize, but usually I wouldn't even get this far.

Thank you! I really appreciate the effort.

![image](https://github.com/user-attachments/assets/3bcf523a-e02a-4b92-904d-4a0fc6d64e01)

**Comment by TibsD** [comment]
How do you install this build? I am waiting to update the HA OS because of this issue so if it would be solved then that would be great.

**Comment by TibsD** [comment]
@piercy will this then let me update the home assistant OS on a regular basis? Now I am holding back on the update HAOS 13.2 and HA Core 2024.11.2 because of the boot issue.

**Comment by sairon** [comment]
@TibsD It should be safe now to upgrade to 14.0.rc1 (and future OS updates) with Kingston eMMC.

**Comment by betriebsanleitung** [comment]
> @TibsD It should be safe now to upgrade to 14.0.rc1 (and future OS updates) with Kingston eMMC.

@sairon, I received the update info for HAOS 14.1 today. There is no mention of #3250 in the release notes for 14.1, while you wrote in your comment above that it is now safe to update HAOS. 
How safe is it to install 14.1 with a Kingston chip on the emmc?

**Comment by sairon** [comment]
Should be fixed via #3665.
